packages/cli/docs/custom-plugins.md

```diff
@@ -0,0 +1,637 @@
+property.
+
+Any plugin is registered in a core config object and can take potential options to configure its behaviour.
+The following example shows where to register the plugin
+
+```typescript
+// code-pushup.config.ts
+import { create } from 'my-plugin';
+
+export default {
+  plugins: [
+    // can return `PluginConfig` or `Promise<PluginConfig>`
+    await create({
+      // plugin options here
+    }),
+  ],
+  categories: [],
+};
+```
+
+## Plugin Structure
+
+Every plugin is defined over a [`PluginConfig`](@TODO).
+
+The plugin config maintains:
+
+- metadata about the plugin
+- metadata about the available audits [`Audit`](@TODO)
+- internal logic producing the plugin output as [`AuditOutputs`](@TODO).
+
+A minimal plugin object maintaining the required fields looks like the following:
+
+**template for a custom plugin**
+
+```typescript
+// my-plugin.ts
+import { AuditOutputs, PluginConfig } from '@code-pushup/models';
+
+export const pluginMeta = {
+  slug: 'my-plugin',
+  title: 'My plugin',
+  // icon name from [vscode-material-icon-theme](https://github.com/PKief/vscode-material-icon-theme/tree/main/icons)
+  icon: 'javascript',
+};
+
+const auditMeta = {
+  slug: 'my-audit',
+  title: 'My audit',
+};
+
+export const audits = [auditMeta];
+
+export type Options = {
+  // your options here
+};
+
+export async function create(options: Options): PluginConfig {
+  return {
+    ...pluginMeta,
+    audits: [audit],
+    runner: runnerFunction,
+  };
+}
+
+function runnerFunction(): AuditOutputs {
+  // Dummy audit output
+  const auditOutput: AuditOutput = {
+    ...auditMeta,
+    score: 0,
+    value: 0,
+  };
+
+  // return dummy data of type `AuditOutputs` as plugin result
+  return [auditOutput];
+}
+```
+
+Execute the CLI with `npx code-pushup collect` and you should the following output:
+
+**stdout of CLI for the above code**
+
+```sh
+Code PushUp Report - @code-pushup/core@x.y.z
+
+My plugin audits
+● My audit                                                            0
+
+Categories
+┌──────────┬───────┬────────┐
+│ Category │ Score │ Audits │
+└──────────┴───────┴────────┘
+
+Made with ❤ by code-pushup.dev
+```
+
+The categories are empty for now. But under the audit listing you can see your plugin title `My plugin`, it's listed
+audit `My audit` and the resulting value `0`.
+
+## Debugging custom plugins
+
+When developing custom plugins you should know a couple of CLI options helpful when debugging.
+
+Following options are helpful in debugging:
+
+- use `--verbose` to get more information printed in the terminal
+- use `--no-progress` to get better readability of logs.  
+  The progressbar would otherwise interfere with your logs and makes them harder to read.
+- use `--onlyPlugin` will restrict the execution of plugins to only the listed ones
+- use `--config` to point to a different config file
+- use `--format=md` to see all information provided by plugin outputs
+
```

Comment:
Since we have an updated CLI README now, I would add a link to the full list of options here.